timer.cpp

```diff
@@ -71,3 +71,17 @@ std::ostream& operator<<(std::ostream& os, const Timer& timer) {
     return os << timer.header << "real = " << std::get<0>(rus) << "s, user = " << std::get<1>(rus)
               << "s, sys = " << std::get<2>(rus) << "s";
 }
+
+StopWatch::StopWatch():
+    start(std::chrono::high_resolution_clock::now())
+{}
+   
+
+TimerGuard::TimerGuard(ElapsedTimeCallback cb):
+    callback(std::move(cb))
+{}
+
+TimerGuard::~TimerGuard() {
+    callback(stopwatch);
```

Comment:
i think you can pass `stopwatch.elapsed()` into 'callback' so that the implementation of 'callback' is independent on 'StopWatch' and 'StopWatch' can be moved to .cpp. All depends on what you'll need...

Reply:
I thought about that but I wanted to have a way to parametrize the unit type (microseconds by default) and this unit type is hold by the `elapsed` function like in:
```cpp
TimerGuard t([&](const StopWatch& sw){ 
        auto micro = sw.elapsed();
        auto milli = sw.elapsed<std::chrono::milliseconds>();
        auto sec = sw.elapsed<std::chrono::seconds>();
        etc....
});
``` 
At first I wanted to hide the `StopWatch` implementation like you said, but I couldn't get something working that would be easily understandable... But my original idea was to have something like :
```cpp
TimerGuard t([&](std::chrono::milliseconds elapsed){ 
        ...
 });
```

Having `StopWatch` exposed within the interface though the callback is the nicest and simplest interface I could get :(